src/timely-util/src/containers.rs

```diff
@@ -9,5 +9,415 @@
+
+    impl<C: Columnar> ContainerBytes for Column<C> {
+        fn from_bytes(bytes: Bytes) -> Self {
+            // Our expectation / hope is that `bytes` is `u64` aligned and sized.
+            // If the alignment is borked, we can relocate. If the size is borked,
+            // not sure what we do in that case.
+            assert_eq!(bytes.len() % 8, 0);
+            if let Ok(_) = bytemuck::try_cast_slice::<_, u64>(&bytes) {
+                Self::Bytes(bytes)
+            } else {
+                println!("Re-locating bytes for alignment reasons");
+                let mut alloc: Vec<u64> = vec![0; bytes.len() / 8];
+                bytemuck::cast_slice_mut(&mut alloc[..]).copy_from_slice(&bytes[..]);
+                Self::Align(alloc.into())
+            }
+        }
+
+        fn length_in_bytes(&self) -> usize {
+            match self {
+                // We'll need one u64 for the length, then the length rounded up to a multiple of 8.
+                Column::Typed(t) => 8 * t.borrow().length_in_words(),
+                Column::Bytes(b) => b.len(),
+                Column::Align(a) => 8 * a.len(),
+            }
+        }
+
+        fn into_bytes<W: ::std::io::Write>(&self, writer: &mut W) {
+            match self {
+                Column::Typed(t) => {
+                    use columnar::Container;
+                    // Columnar data is serialized as a sequence of `u64` values, with each `[u8]` slice
+                    // serialize as first its length in bytes, and then as many `u64` values as needed.
+                    // Padding should be added, but only for alignment; no specific values are required.
+                    for (align, bytes) in t.borrow().as_bytes() {
+                        assert!(align <= 8);
+                        let length: u64 = bytes.len().try_into().unwrap();
+                        writer
+                            .write_all(bytemuck::cast_slice(std::slice::from_ref(&length)))
+                            .unwrap();
+                        writer.write_all(bytes).unwrap();
+                        let padding: usize = ((8 - (length % 8)) % 8).try_into().unwrap();
+                        writer.write_all(&[0; 8][..padding]).unwrap();
+                    }
+                }
+                Column::Bytes(b) => writer.write_all(b).unwrap(),
+                Column::Align(a) => writer.write_all(bytemuck::cast_slice(a)).unwrap(),
+            }
+        }
+    }
+}
+
+pub use builder::ColumnBuilder;
+mod builder {
+    use std::collections::VecDeque;
+
+    use columnar::{AsBytes, Clear, Columnar, Len, Push};
+    use timely::container::PushInto;
+    use timely::container::{ContainerBuilder, LengthPreservingContainerBuilder};
+
+    use super::Column;
+
+    /// A container builder for `Column<C>`.
+    pub struct ColumnBuilder<C: Columnar> {
+        /// Container that we're writing to.
+        current: C::Container,
+        /// Empty allocation.
+        empty: Option<Column<C>>,
```

Comment:
This field is only used so we can give out references to finished containers instead of owned types, right? Could we make this clearer in the name and/or comment of the field? The `Chunker` below does a good job of this.